File: Bag.py
from Pennant import Pennant

class Bag:
    data = None
    def __init__(self, n = 0):
        if n == 0:
            self.data = [None]
        else:
            self.data = [None] * n

    def addElement(self, pennant):
        success = False
        i = 0
        current = pennant
        #print("Node to add: ", pennant.masterNodeValue)
        while success == False:
            
            if i >= len(self.data):
                break
            #print("--")
            #current.printPennant()
            if self.data[i] == None:
                self.data[i] = current
                success = True
                #print(self.data[i].printPennant())
                break

            else:
                self.data[i].joinPennant(current)
                current = self.data[i]
                self.data[i] = None
            
            
            i += 1
            
        if success == False:
            self.data.append(current)

        #current.printPennant()

    def split(self):
        bag_left = Bag(len(self.data) - 1)
        bag_right = Bag(len(self.data) - 1)

        #print(len(self.data))
        #self.data = self.data[::-1]
        print("Splitting")

        
        i = len(self.data) - 1

        while i > 0:
            if self.data[i] == None:
                i -= 1

            else:
                p_left, p_right = self.data[i].splitPennant()
                bag_left.data[i - 1] = p_left
                bag_right.data[i - 1] = p_right
                i -= 1
        
        if self.data[0] != None:
            bag_right.addElement(self.data[0])

        
        self = bag_left

        return bag_right



    def printBag(self):
        #ans = []
        for i in self.data:
            if i == None:
                print("None_")
            else:
                print("Node: ", i.masterNodeValue)
                i.printPennant()
                
        
        
if __name__ == "__main__":
    b = Bag()
    r = [1, 2, 3, 4, 5]
    for i in r:
        b.addElement(Pennant(i))

    b1 = b.split()


    b.printBag()
    print("------")
    b1.printBag()

File: Bag.java
import java.util.ArrayList;

public class Bag {

    private int size;
    private ArrayList<Pennant> items;

    public Bag() {
        this.size = 0;
        this.items = new ArrayList<Pennant>();
        this.items.add(null);

    }

    public Bag(int n) {
        this.size = 0;
        this.items = new ArrayList<Pennant>();
        for(int i = 0; i < n; i++) {
            this.items.add(null);
        }
    }

    public void addPennant(int position, Pennant p) {
        this.items.set(position, p);
    }

    public ArrayList<Pennant> getItems() {
        return this.items;
    }

    public void addElement(Pennant p) {
        boolean success = false;

        int i = 0;
        while(success == false) {
            if(i >= this.items.size()) {
                break;
            }

            if(this.items.get(i) == null) {
                this.items.set(i, p);
                success = true;
                break;
            }
            else {
                this.items.get(i).joinPennant(p);
                p = this.items.get(i);
                this.items.set(i, null);
                i++;
            }
        }

        if(success == false) {
            this.items.add(p);
        }
    }

    public Bag splitBag() {
        Bag bag_left = new Bag(this.items.size() - 1);
        Bag bag_right = new Bag(this.items.size() - 1);

        for(int i = this.items.size() - 1; i > 0; i--) {
            if(this.items.get(i) == null) {
                continue;
            }
            Pennant p_right = this.items.get(i).splitPennant();
            bag_left.addPennant(i - 1, this.items.get(i));
            bag_right.addPennant(i - 1, p_right);
        }

        if(this.items.get(0) != null) {
            bag_right.addElement(this.items.get(0));
        }

        this.items = bag_left.getItems();

        return bag_right;
    }


    public void printBag() {
        for(int i = 0; i < this.items.size(); i++) {
            if(this.items.get(i) == null) {
                System.out.println("No Element");
            }
            else {
                this.items.get(i).printPennant();
            }
            System.out.println("------");
        }
    }


    public static void main(String[] args) {
        System.out.println("Begin");

        Pennant p1 = new  Pennant(1);
        Pennant p2 = new Pennant(2);
        Pennant p3 = new Pennant(3);
        Pennant p4 = new Pennant(4);
        Pennant p5 = new Pennant(5);

        Bag b1 = new Bag();
        b1.addElement(p1);
        b1.addElement(p2);
        b1.addElement(p3);
        b1.addElement(p4);
        b1.addElement(p5);

        //b1.printBag();

        b1.splitBag();
        b1.printBag();

    }
}